src/pycsou/opt/solver/lsqr.py

```diff
@@ -0,0 +1,319 @@
+        equations and sparse least squares", Dissertation,
+        https://web.stanford.edu/group/SOL/software/lsqr/lsqr-toms82a.pdf
+    """
+
+    def __init__(
+        self,
+        A: pyco.PosDefOp,
+        *,
+        damp: float = 0.0,
+        atol: float = 1e-06,
+        btol: float = 1e-06,
+        conlim: float = 1e08,
+        iter_lim: typ.Optional[int] = None,
+        eps: float = np.finfo(np.float64).eps,
+        folder: typ.Optional[pyct.PathLike] = None,
+        exist_ok: bool = False,
+        writeback_rate: typ.Optional[int] = None,
+        verbosity: int = 1,
+        show_progress: bool = True,
+        log_var: pyct.VarName = ("x",),
+    ):
+        super().__init__(
+            folder=folder,
+            exist_ok=exist_ok,
+            writeback_rate=writeback_rate,
+            verbosity=verbosity,
+            show_progress=show_progress,
+            log_var=log_var,
+        )
+
+        self._A = A
+        self._atol = atol
+        self._btol = btol
+        self._ctol = (1.0 / conlim) if (conlim > 0) else 0.0
+        self._iter_lim = (2 * A.shape[1]) if (iter_lim is None) else iter_lim
+        self._dampsq, self._damp = damp**2, damp
+        self._eps = eps
+
+    @pycrt.enforce_precision(i=["b", "x0"], allow_None=True)
+    def m_init(
+        self,
+        b: pyct.NDArray,
+        x0: typ.Optional[pyct.NDArray] = None,
+    ):
+        xp = pycu.get_array_module(b)
+
+        b = xp.atleast_1d(b)
+        if b.ndim > 1:
+            b = b.squeeze()
+        m, n = self._A.shape
+
+        mst = self._mstate
+        mst["normA"] = mst["condA"] = mst["res2"] = mst["xnorm"] = mst["xxnorm"] = 0.0
+        mst["ddnorm"] = mst["z"] = mst["sn2"] = 0.0
+        mst["cs2"] = -1.0
+
+        u = b
+        bnorm = xp.linalg.norm(b)
+
+        if x0 is None:
+            x = xp.zeros(n)
+            beta = bnorm.copy()
+        else:
+            x = xp.asarray(x0)
+            u = u - self._A.apply(x)
+            beta = xp.linalg.norm(u)
+
+        if beta > 0:
+            u = (1 / beta) * u
+            v = self._A.T.apply(u)
+            alpha = xp.linalg.norm(v)
+        else:
+            v = x.copy()
+            alpha = 0
+
+        if alpha > 0:
+            v = (1 / alpha) * v
+
+        arnorm = alpha * beta
+        if arnorm == 0:
+            mst["trivial"] = True
+            if b.ndim == 1:
+                mst["x"] = xp.zeros(n)
+            elif b.ndim == 2:
+                mst["x"] = xp.zeros((b.shape[0], n))
+            return
+
+        mst["x"] = x
+        mst["u"], mst["v"] = u, v
+        mst["w"] = v.copy()
+        mst["rhobar"] = mst["alpha"] = alpha
+        mst["phibar"] = mst["rnorm"] = mst["normr1"] = mst["normr2"] = beta
+        mst["bnorm"] = bnorm
+        mst["trivial"] = False
+        mst["test1"], mst["test2"] = 1.0, alpha / beta
+        mst["test3"] = mst["t1"] = mst["rtol"] = None
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
```

Comment:
would it be interesting to use a base class for LSMR and LSQR which defines the common methods, _sym_ortho, solution, input handling, etc..?

Reply:
Haven't had a look at the code, but it could be a good idea indeed: the two algorithms are very similar and if there is redundancy, let's centralise it into a private class (maybe even an abstract one as _PDS).